dartagnan/src/main/java/com/dat3m/dartagnan/program/Thread.java

```diff
@@ -56,6 +56,10 @@ public Set<Thread> getSyncSet() {
         return syncSet.get();
     }
 
+    public void setScopeHierarchy(ScopeHierarchy scopeHierarchy) {
+        this.scopeHierarchy = Optional.of(scopeHierarchy);
+    }
+
```

Comment:
Related to a previous comment: do we really need to explicitly add the hierarchy to a thread that did not have one? This sounds to me like we are doing things wrong: either the hierarchy is set when creating the thread, of we don't need a hierarcy at all.

Reply:
The scopeHierarchy is changed back to final now after reordering the visit of thread declarator and variable declarator